r2-core/src/main/java/com/linkedin/r2/transport/http/client/SmoothRateLimiter.java

```diff
@@ -272,6 +272,13 @@ public void loop()
 
       if (_permitAvailableCount > 0)
       {
+        if (!_wasDelayed)
+        {
+          _wasDelayed = true;
+          _scheduler.schedule(this::loop, _executionTracker.getNextExecutionDelay(_rate), TimeUnit.MILLISECONDS);
+          return;
+        }
+        _wasDelayed = false;
```

Comment:
And we don't need to set it here right?

Reply:
There might be a more readable way of doing this, totally open to suggestions. The whole purpose of lines 275 - 281 is to stop prematurely every other iteration of the loop, taking care to schedule the next loop iteration after the delay. This works because the delay is always set such that there's time enough in the rest of the period to honor all the permits.